README.md
[![Published on NPM](https://img.shields.io/npm/v/@api-components/api-body-document.svg)](https://www.npmjs.com/package/@api-components/api-body-document)

[![Build Status](https://travis-ci.org/advanced-rest-client/api-body-document.svg?branch=stage)](https://travis-ci.org/advanced-rest-client/api-body-document)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/advanced-rest-client/api-body-document)

## &lt;api-body-document&gt;

A component to render HTTP method body documentation based on AMF model generated from API spec file.

## Version compatibility

This version only works with AMF model version 2 (AMF parser >= 4.0.0).
For compatibility with previous model version use `3.x.x` version of the component.

## Styling

`<api-body-document>` provides the following custom properties and mixins for styling:

Custom property | Description | Default
----------------|-------------|----------
`--api-body-document-title-border-color` | Border color of the section title | `#e5e5e5`
`--api-body-document-toggle-view-color` | Color of the toggle view button | `--arc-toggle-view-icon-color` or `rgba(0, 0, 0, 0.74)`
`--api-body-document-toggle-view-hover-color` | Color of the toggle view button when hovered | `var(--arc-toggle-view-icon-hover-color` or `rgba(0, 0, 0, 0.88)`
`--api-body-document-description-color` | Color of the type description | `rgba(0, 0, 0, 0.74)`
`--api-body-document-media-button-background-color` | Selection color of the media type selector | `#CDDC39`
`--api-body-document-examples-title-color` | Color of examples section title | ``
`--api-body-document-examples-border-color` | Example section border color | `transparent`
`--code-background-color` | Background color of the examples section | ``
`--api-body-document-media-type-label-font-weight` | Font weight of the media type label (when selection is not available) | `500`
`--api-body-document-title-narrow-font-size`  |   |  `initial`,
`--api-body-document-code-color`  |   |  `initial`
`--api-body-document-any-info-font-size`  |   |  `16px`
`--api-body-document-any-info-font-weight`  |   |  `500`

### API components

This components is a part of [API components ecosystem](https://elements.advancedrestclient.com/)

## Usage

### Installation
```
npm install --save @api-components/api-body-document
```

### In an html file

```html
<html>
  <head>
    <script type="module">
      import '@api-components/api-body-document/api-body-document.js';
    </script>
  </head>
  <body>
    <api-body-document></api-body-document>
  </body>
</html>
```

### In a LitElement

```js
import { LitElement, html } from 'lit-element';
import '@api-components/api-body-document/api-body-document.js';

class SampleElement extends PolymerElement {
  render() {
    return html`
    <api-body-document .amf="${this.amf}"></api-body-document>
    `;
  }
}
customElements.define('sample-element', SampleElement);
```

## Development

```sh
git clone https://github.com/advanced-rest-client/api-body-document
cd api-body-document
npm install
```

### Running the demo locally

```sh
npm start
```

### Running the tests
```sh
npm test
```


CHANGELOG.md
<a name="2.0.14"></a>
## 2.0.14 (2019-01-17)


### Breaking

* Updating AMF data model ([2286bb40557de75c3679d5d44468aab4d7253cb5](https://github.com/advanced-rest-client/api-body-document/commit/2286bb40557de75c3679d5d44468aab4d7253cb5))

### Docs

* Updated documentation ([1455ffebbc7e286bde47f3cc9573b1427614ca73](https://github.com/advanced-rest-client/api-body-document/commit/1455ffebbc7e286bde47f3cc9573b1427614ca73))
* Updating documentation after styles API change ([eeeeb00475ba23af9475e72eb081d3534376a961](https://github.com/advanced-rest-client/api-body-document/commit/eeeeb00475ba23af9475e72eb081d3534376a961))
* Updating documentation. ([6bb206b8cc9b7053c051b812666f6e731d3da041](https://github.com/advanced-rest-client/api-body-document/commit/6bb206b8cc9b7053c051b812666f6e731d3da041))

### Fix

* Fixing examples rendering ([948a596a9aae7f8ca88e7e4900fa0b19256ca363](https://github.com/advanced-rest-client/api-body-document/commit/948a596a9aae7f8ca88e7e4900fa0b19256ca363))

### Update

* Adding array example computation ([aaaae1869e586503b2baf4bdff24e2dadbca203e](https://github.com/advanced-rest-client/api-body-document/commit/aaaae1869e586503b2baf4bdff24e2dadbca203e))
* Adding info message for "any" data type ([b2d16bfde693844659530cf0861881b67ec6f81d](https://github.com/advanced-rest-client/api-body-document/commit/b2d16bfde693844659530cf0861881b67ec6f81d))
* Adding support for compact json-ld model ([c48372025db7e3d399ab4d91947607a662b160de](https://github.com/advanced-rest-client/api-body-document/commit/c48372025db7e3d399ab4d91947607a662b160de))
* Making whole section title area clickable ([11db259fa8ac2ea0ce239b7bf501d854681d810b](https://github.com/advanced-rest-client/api-body-document/commit/11db259fa8ac2ea0ce239b7bf501d854681d810b))
* Updated Travis configuration to connect to Sauce Labs. ([af0acccd2dda343b59fd4ec59ad49cf6930167e5](https://github.com/advanced-rest-client/api-body-document/commit/af0acccd2dda343b59fd4ec59ad49cf6930167e5))
* Updating component to use compact data model ([5f359a191a0df180f45fd48aeb11788ef0fd0dda](https://github.com/advanced-rest-client/api-body-document/commit/5f359a191a0df180f45fd48aeb11788ef0fd0dda))
* Updating styling API - adding `narrow` attribute ([41fdb940a62f27ceee0a86f5e4fd3815b5beaedd](https://github.com/advanced-rest-client/api-body-document/commit/41fdb940a62f27ceee0a86f5e4fd3815b5beaedd))



<a name="2.0.13"></a>
## 2.0.13 (2018-12-28)


### Breaking

* Updating AMF data model ([2286bb40557de75c3679d5d44468aab4d7253cb5](https://github.com/advanced-rest-client/api-body-document/commit/2286bb40557de75c3679d5d44468aab4d7253cb5))

### Docs

* Updated documentation ([1455ffebbc7e286bde47f3cc9573b1427614ca73](https://github.com/advanced-rest-client/api-body-document/commit/1455ffebbc7e286bde47f3cc9573b1427614ca73))
* Updating documentation after styles API change ([eeeeb00475ba23af9475e72eb081d3534376a961](https://github.com/advanced-rest-client/api-body-document/commit/eeeeb00475ba23af9475e72eb081d3534376a961))
* Updating documentation. ([6bb206b8cc9b7053c051b812666f6e731d3da041](https://github.com/advanced-rest-client/api-body-document/commit/6bb206b8cc9b7053c051b812666f6e731d3da041))

### Fix

* Fixing examples rendering ([948a596a9aae7f8ca88e7e4900fa0b19256ca363](https://github.com/advanced-rest-client/api-body-document/commit/948a596a9aae7f8ca88e7e4900fa0b19256ca363))

### Update

* Adding array example computation ([aaaae1869e586503b2baf4bdff24e2dadbca203e](https://github.com/advanced-rest-client/api-body-document/commit/aaaae1869e586503b2baf4bdff24e2dadbca203e))
* Adding info message for "any" data type ([b2d16bfde693844659530cf0861881b67ec6f81d](https://github.com/advanced-rest-client/api-body-document/commit/b2d16bfde693844659530cf0861881b67ec6f81d))
* Adding support for compact json-ld model ([c48372025db7e3d399ab4d91947607a662b160de](https://github.com/advanced-rest-client/api-body-document/commit/c48372025db7e3d399ab4d91947607a662b160de))
* Making whole section title area clickable ([11db259fa8ac2ea0ce239b7bf501d854681d810b](https://github.com/advanced-rest-client/api-body-document/commit/11db259fa8ac2ea0ce239b7bf501d854681d810b))
* Updated Travis configuration to connect to Sauce Labs. ([af0acccd2dda343b59fd4ec59ad49cf6930167e5](https://github.com/advanced-rest-client/api-body-document/commit/af0acccd2dda343b59fd4ec59ad49cf6930167e5))
* Updating component to use compact data model ([5f359a191a0df180f45fd48aeb11788ef0fd0dda](https://github.com/advanced-rest-client/api-body-document/commit/5f359a191a0df180f45fd48aeb11788ef0fd0dda))
* Updating styling API - adding `narrow` attribute ([41fdb940a62f27ceee0a86f5e4fd3815b5beaedd](https://github.com/advanced-rest-client/api-body-document/commit/41fdb940a62f27ceee0a86f5e4fd3815b5beaedd))



<a name="2.0.12"></a>
## 2.0.12 (2018-10-30)


### Breaking

* Updating AMF data model ([2286bb40557de75c3679d5d44468aab4d7253cb5](https://github.com/advanced-rest-client/api-body-document/commit/2286bb40557de75c3679d5d44468aab4d7253cb5))

### Docs

* Updated documentation ([1455ffebbc7e286bde47f3cc9573b1427614ca73](https://github.com/advanced-rest-client/api-body-document/commit/1455ffebbc7e286bde47f3cc9573b1427614ca73))
* Updating documentation after styles API change ([eeeeb00475ba23af9475e72eb081d3534376a961](https://github.com/advanced-rest-client/api-body-document/commit/eeeeb00475ba23af9475e72eb081d3534376a961))
* Updating documentation. ([6bb206b8cc9b7053c051b812666f6e731d3da041](https://github.com/advanced-rest-client/api-body-document/commit/6bb206b8cc9b7053c051b812666f6e731d3da041))

### Fix

* Fixing examples rendering ([948a596a9aae7f8ca88e7e4900fa0b19256ca363](https://github.com/advanced-rest-client/api-body-document/commit/948a596a9aae7f8ca88e7e4900fa0b19256ca363))

### Update

* Adding array example computation ([aaaae1869e586503b2baf4bdff24e2dadbca203e](https://github.com/advanced-rest-client/api-body-document/commit/aaaae1869e586503b2baf4bdff24e2dadbca203e))

src/Styles.js
import { css } from 'lit-element';

export default css`
:host {
  display: block;
  font-size: var(--arc-font-body1-font-size);
  font-weight: var(--arc-font-body1-font-weight);
  line-height: var(--arc-font-body1-line-height);
}

[hidden] {
  display: none !important;
}

.section-title-area {
  display: flex;
  flex-direction: row;
  align-items: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-bottom: 1px var(--api-body-document-title-border-color, #e5e5e5) solid;
  transition: border-bottom-color 0.15s ease-in-out;
}

.section-title-area[opened] {
  border-bottom-color: transparent;
}

.section-title-area .table-title {
  flex: 1;
  flex-basis: 0.000000001px;
  font-size: var(--api-body-document-title-narrow-font-size, initial);
}

.toggle-button {
  outline: none;
}

.toggle-icon {
  margin-left: 8px;
  transform: rotateZ(0deg);
  transition: transform 0.3s ease-in-out;
}

.toggle-icon.opened {
  transform: rotateZ(-180deg);
}

.table-title {
  font-size: var(--arc-font-subhead-font-size);
  font-weight: var(--arc-font-subhead-font-weight);
  line-height: var(--arc-font-subhead-line-height);
}

:host([narrow]) .table-title {
  font-size: var(--api-body-document-title-narrow-font-size, initial);
}

.type-title {
  font-size: var(--arc-font-body2-font-size);
  font-weight: var(--arc-font-body2-font-weight);
  line-height: var(--arc-font-body2-line-height);
}

.body-name {
  font-weight: var(--api-body-document-any-info-font-weight, 500);
  font-size: 1.1rem;
}

anypoint-button[active] {
  background-color: var(--api-body-document-media-button-background-color, #CDDC39);
}

.media-type-selector {
  margin: 20px 0;
}

.markdown-html {
  margin-bottom: 28px;
  margin-top: 28px;
  color: var(--api-body-document-description-color, rgba(0, 0, 0, 0.74));
}

.markdown-html[data-with-title] {
  margin-top: 0;
}

api-schema-document {
  background-color: var(--code-background-color);
  padding: 8px;
  color: var(--api-body-document-code-color, initial);
  word-break: break-all;
}

.media-type-label {
  font-weight: var(--api-body-document-media-type-label-font-weight, 500);
  margin-left: 8px;
}

.media-toggle {
  outline: none;
  color: var(--api-body-document-toggle-view-color, var(--arc-toggle-view-icon-color, rgba(0, 0, 0, 0.74)));
}

.any-info,
.any-info-description {
  color: var(--api-body-document-description-color, rgba(0, 0, 0, 0.74));
}

.any-info {
  font-size: var(--api-body-document-any-info-font-size, 16px);
}

arc-marked {
  background-color: transparent;
  padding: 0px;
}

.icon {
  display: block;
  width: 24px;
  height: 24px;
  fill: currentColor;
}
`;
